include/nbl/asset/filters/CFlattenRegionsStreamHashImageFilter.h

```diff
@@ -0,0 +1,220 @@
+// Copyright (C) 2018-2024 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef __NBL_ASSET_C_FLATTEN_REGIONS_STREAM_HASH_IMAGE_FILTER_H_INCLUDED__
+#define __NBL_ASSET_C_FLATTEN_REGIONS_STREAM_HASH_IMAGE_FILTER_H_INCLUDED__
+
+#include <type_traits>
+
+#include "nbl/core/declarations.h"
+
+#include "nbl/asset/filters/CMatchedSizeInOutImageFilterCommon.h"
+#include "nbl/asset/filters/CFlattenRegionsImageFilter.h"
+#include "blake/c/blake3.h"
+
+namespace nbl
+{
+namespace asset
+{
+class CFlattenRegionsStreamHashImageFilter : public CMatchedSizeInOutImageFilterCommon
+{
+	public:
+		virtual ~CFlattenRegionsStreamHashImageFilter() {}
+
+		struct ScratchMemory
+		{
+			core::smart_refctd_ptr<asset::ICPUImage> flatten; // for flattening input regions & prefilling with 0-value texels not covered by regions 
+			core::smart_refctd_ptr<asset::ICPUBuffer> heap; // for storing hashes, single hash is obtained from given miplevel & layer, full hash for an image is a hash of this stack
+		};
+		
+		class CState : public IImageFilter::IState
+		{
+			public:
+				virtual ~CState() {}
+
+				using hash_t = std::array<size_t, 4u>;
+
+				const ICPUImage*					inImage = nullptr;
+				hash_t								outHash = { {0} };
+				ScratchMemory						scratchMemory;
+		};
+
+		using state_type = CState;
+
+		static inline ScratchMemory allocateScratchMemory(const asset::ICPUImage* inImage)
+		{
+			ScratchMemory scratch;
+			scratch.flatten = asset::IAsset::castDown<asset::ICPUImage>(inImage->clone());
```

Comment:
you need to do a level 0 clone, otherwise you get a buffer cope as well

Reply:
the copy was intended, but yes it's a dumb copy

Reply:
done